Changes in onnxmlir package

chentong319 opened the pull request:
There are two changes in this PR:

1. Add the support for optional dependence package, docker or podman 

- if you want to use docker package, install with `pip install -e path/onnxmlir[docker]`
- if you want to use podman package, install with `pip install -e path/onnxmlir[podman]`

2. Add the support for different python version. 

-  The compiled PyRuntimeC*.so will be copied to onnxmlir/src/onnxmlir/libs/major_minor directory. Manual work is still needed to integrate different version together
- When the package onnxmlir is used. The current python version is checked and the corresponding  PyRuntimeC*.so is imported. 

3. Make ONNXRUNTIME interface as special case of run_ort(), while use the PyRuntime interface as the default for run(). Rewrite the test cases.
4. More tests and fixed some bugs

- Create a git repo for the package: https://github.ibm.com/chentong/onnx-mlir-python-driver
- Test natively on z16 and z17
- Test inside onnx-mlir container

AlexandreEichenberger commented on src/Runtime/python/onnxmlir/pyproject.toml:
```diff
@@ -16,11 +16,15 @@ classifiers = [
-dependences = [
-    "numpy",
-    "docker",
+
```
Somewhere there should be something that allow the install package to say "sorry, not supported on platform X". Ok to be zLinux only for now.

AlexandreEichenberger commented on src/Runtime/python/onnxmlir/tests/use_compiler_container.py:
```diff
@@ -0,0 +1,21 @@
+sess = onnxmlir.InferenceSession(
+    "test_add.onnx",
+    compile_args="-O3 --parallel",
+    container_engine="docker",
```
Docker vs Podman. If there was a way to "query" which package was installed (as we had to do a `install onnxmlir[docker]` vs `onnxmlir[podman]`, that would be one less option that seems redundant.

AlexandreEichenberger reviewed (commented):
LGTM, fantastic to be so close.

Will approve once the (python)
```
# SPDX-License-Identifier: Apache-2.0
```

or (c)
```
/*
 * SPDX-License-Identifier: Apache-2.0
 */
```
or (md files)
```
<!--- SPDX-License-Identifier: Apache-2.0 -->
```

markers are added in the respective files.

chentong319 commented on docs/BuildPyRuntimeLit.md:
````diff
@@ -2,38 +2,51 @@
+```
+cd path/to/existing/onnx-mlir
+mkdir build-light
+cd build-ligth
````
Fixed

chentong319 commented on src/Runtime/python/onnxmlir/README.md:
````diff
@@ -1,68 +1,57 @@
 ```
-
-With onnxmlir package, onnx-mlir can be used to replace onnxrutnime as follows:
+git clone git@github.ibm.com:chentong/onnx-mlir-python-driver.git
````
Yes, we will decide this issue later.

chentong319 commented on docs/BuildPyRuntimeLit.md:
```diff
@@ -2,38 +2,51 @@
-All necessary dependence, such as llvm_project and onnx-mlir compiler is removed. The purpose is to easily build the python driver for the model execution on 
-different systems. Currently, only the OMTenserUtils (src/Runtime), Python driver (src/Runtime/python), third_party/onnx and third_party/pybind11 are built.
-
+onnx-mlir compiler can compile an ONNX model into a shared library (.so), and also provide a python driver to run the generated shared library through python code. That is PyRuntimeC. Originally, the PyRuntimeC is built with the onnx-mlir compiler, and consequently need the build of llvm_project. The PyRuntimeC lit is a different way to build the python d PyRuntime (src/Runtime/python) that does  ...
```
Fixed

chentong319 commented:
@AlexandreEichenberger @tungld Thanks for reviewing. I addressed most of your concerns. It is not settled in this PR is where to store the package. The pre-compiled libs are the result of build for different system. We may need a separate git for the package.
